### PR TITLE
Erc20 components 

### PR DESCRIPTION
**Pull request by notV4l**
WIP

### CODE REVIEW

File: token/src/components/token/erc20/erc20_burnable.cairo [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC20Balance: erc20_balance_comp::HasComponent<TContractState>,
+        impl ERC20Metadata: erc20_metadata_comp::HasComponent<TContractState>,
+        +Drop<TContractState>,
+    > of InternalTrait<TContractState> {
+        fn _burn(ref self: ComponentState<TContractState>, account: ContractAddress, amount: u256) {
```
Comment (tarrencev) [inline token/src/components/token/erc20/erc20_burnable.cairo:34]:
```suggestion
        fn burn(ref self: ComponentState<TContractState>, account: ContractAddress, amount: u256) {
```

File: token/src/components/tests/mocks/erc20/erc20_allowance_mock.cairo [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#[dojo::contract]
+mod ERC20AllowanceMock {
```
[the remark this answers is not in the document]
Comment (notV4l) [reply token/src/components/tests/mocks/erc20/erc20_allowance_mock.cairo:2]: OZ uses CamelCase, starkware's tests files uses snake_case ..

[the remark this answers is not in the document]
Comment (notV4l) [reply token/src/components/tests/mocks/erc20/erc20_allowance_mock.cairo]: updated to snake_case

File: token/src/components/token/erc20/erc20_allowance.cairo [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+        // use in transfer_from
+        fn _spend_allowance(
+            ref self: ComponentState<TContractState>,
+            owner: ContractAddress,
+            spender: ContractAddress,
+            amount: u256
+        ) {
+            let current_allowance = self.get_allowance(owner, spender).amount;
+            if current_allowance != BoundedInt::max() {
```
[the remark this answers is not in the document]
Comment (notV4l) [reply token/src/components/token/erc20/erc20_allowance.cairo:204]: owner allowance get consumed when an allowed spender transfer_from owner

**Review by Bal7hazar — APPROVED** [review]
Ok I am fine with the current version.

I am a bit worried about to have our own codebase regarding tokens instead of having dependency with OZ, but I don't see any other way to have it with the current Cairo limitations (no inheritance mechanism mainly).

So we will have to check sometimes the OZ repo to see if there are some hotfix and update our codebase accordingly.

**Pull request merged**